### DISCUSSION TITLE
Nested logical operators create incorrect database queries

### DISCUSSION THREAD

**Discussion by nikitaDanilenko** [directus/directus#13203] [Feature Requests]
Nested logical operators on nested relations create a wrong database query when the nesting takes place at the lowest level, but works correctly when nested at the highest level.

### To Reproduce

Create the following three tables:

```sql
create table top(
  id int primary key,
  name text not null
);

create table center(
  id int primary key,
  description text not null, 
  top_id int not null references top(id)
);

create table bottom(
  id int primary key,
  label text not null,
  center_id int not null references center(id)
);
```

and make the collections visible via the Directus administration interface. Then run the following query:

```graphql
query {
  bottom(
    filter: {
      center_id: {
        top_id: {
          _and: [
            {id: {_eq: 42}},
            {name: {_contains: "a"}}
          ]
        }
      }
    }
  )
  { id }
}
```

The response is

```json
{
  "errors": [
    {
      "message": "select \"bottom\".\"id\" from \"bottom\" left join \"center\" as \"gpdyu\" on \"bottom\".\"center_id\" = \"gpdyu\".\"id\" left join \"top\" as \"cvrqz\" on \"gpdyu\".\"top_id\" = \"cvrqz\".\"id\" where (\"gpdyu\".\"top_id\" = $1) order by \"bottom\".\"id\" asc limit $2 - invalid input syntax for integer: \"{\"{\\\"id\\\":{\\\"_eq\\\":\\\"42\\\"}}\",\"{\\\"name\\\":{\\\"_contains\\\":\\\"fish\\\"}}\"}\"",
      "locations": [
        {
          "line": 2,
          "column": 3
        }
      ],
      "path": [
        "bottom"
      ]
    }
  ],
  "data": {
    "bottom": null
  }
}
```

Lifting the `_and` to the highest level yields the correct result:

```graphql
query {
  bottom(
    filter: {
      _and: [
        { center_id: { top_id: { id: { _eq: 42 } } } }
        { center_id: { top_id: { name: { _contains: "a" } } } }
      ]
    }
  ) {
    id
  }
}
```

correctly returns

```json
{
  "data": {
    "bottom": []
  }
}
```

If my understanding is correct, the above queries should be equivalent, because they both express that I wish to fetch those `bottom` values, whose `center_id` has a `top_id` with `id = 42` and a `name` containing an `a`. Judging by the database error it seems that in the first expression the expression in the inner `_and` is mistakenly taken as the entire value to search for in `"uczhw"."top_id"`.

**Comment by azrikahar** [comment]
Moving this to discussion as it's technically not a bug (working as intended and as documented), and I've updated the title from "Aggregated Filters" to "Nested Logical Operators" to make it clearer, since we also do actually have aggregate queries as well. Do let me know if that's not what you meant 👍

---

With that said, I can definitely see how your suggestion may feel more intuitive! Unfortunately it's just not how the current logic works, as [Logical Operators](https://docs.directus.io/configuration/filter-rules/#logical-operators) should be top level (but can be nested among themselves), whereas nested operators within fields can only be [Filter Operators](https://docs.directus.io/configuration/filter-rules/#filter-operators), at least as of now.

**Reply by nikitaDanilenko** [reply]
Thanks for the quick feedback and more concise title 👍!

Is it possible that the condition "operators may only nest on top level" could be stated more explicitly together with the rules for the logical and filter operators? I do not mind the condition itself, although dropping it may improve the DRY factor. My main concern is that I can write a valid query on the GraphQL level, but get an invalid resulting database query.

**Reply by azrikahar** [reply]
> Is it possible that the condition "operators may only nest on top level" could be stated more explicitly together with the rules for the logical and filter operators? 

[The documented syntax](https://docs.directus.io/configuration/filter-rules/#syntax) does mention that the **operator** can only be Filter Operators, but perhaps it can be highlighted a lil more in case it's not prominent enough. cc @erondpowell for his thoughts on this.

> My main concern is that I can write a valid query on the GraphQL level, but get an invalid resulting database query.

I believe it actually shouldn't be valid, at least that's what I can observe when there's introspection and autocomplete as seen below. Is it showing up in your setup which may cause this confusion? If it is, would you mind sharing a screenshot or short clip?

(Notice how the suggestions in `dob` (a field) does only show the filter operators, but not logical operators like `_and`/`_or`. and also only filter operators available to date field type)

![chrome_0hRLXqyVW4](https://user-images.githubusercontent.com/42867097/167613409-c8746044-5e59-4a13-92af-e7c95a0a0f74.gif)

**Reply by nikitaDanilenko** [reply]
> [The documented syntax](https://docs.directus.io/configuration/filter-rules/#syntax) does mention that the operator can only be Filter Operators

You are right, of course! The hint is very subtle, but it is definitely there!

> I believe it actually shouldn't be valid

That would be a great solution as well. However, with Insomnia I get the following suggestions

![filter](https://user-images.githubusercontent.com/4136596/167626706-3c7e50bb-a868-4d79-8a39-4887253dd331.gif)

Similarly, I have used the [Caliban](https://github.com/ghostdogpr/caliban) library to create Scala code for the generated GraphQL schema, and the resulting code allows logical operators at the lowest level as well.

**Reply by azrikahar** [reply]
Ahh good one. Seems like they are showing for junction fields (nested fields still won't show and/or like when you were typing `{name: {_contains: "A"}}`). 🤔

**Reply by erondpowell** [reply]
> [The documented syntax](https://docs.directus.io/configuration/filter-rules/#syntax) does mention that the **operator** can only be Filter Operators, but perhaps it can be highlighted a lil more in case it's not prominent enough. cc @erondpowell for his thoughts on this.

Noted. PR made to update and clarify this.

**Comment by rijkvanzanten** [comment]
Heya! Thank you for taking the time to submit this request! 

It has been over **90 days**, and this discussion has not received at least **15 votes** from the community. This means that we don't feel like there's enough community interest to warrant further R&D into this topic at this time. :ice_cube:

This request will now be closed to keep our discussions tidy. Please reach out if you have any questions!

_For more information, see our [Feature Request Process](https://docs.directus.io/contributing/feature-request-process.html)._
